src/sdam/state/mod.rs

```diff
@@ -2,161 +2,302 @@ pub(super) mod server;
+    ) {
+        // If we encounter certain errors, we must update the topology as per the
+        // SDAM spec.
+        if error.is_non_timeout_network_error() {
+            self.mark_server_as_unknown(error, server.address.clone())
+                .await;
+            server.clear_connection_pool();
+        } else if error.is_recovering() || error.is_not_master() {
+            self.mark_server_as_unknown(error.clone(), server.address.clone())
+                .await;
+
+            self.common.message_manager.request_topology_check();
+
+            let wire_version = conn
+                .stream_description()
+                .map(|sd| sd.max_wire_version)
+                .ok()
+                .and_then(std::convert::identity)
+                .unwrap_or(0);
+
+            // in 4.2+, we only clear connection pool if we've received a
+            // "node is shutting down" error. Otherwise, we always clear the pool.
+            if wire_version < 8 || error.is_shutting_down() {
+                server.clear_connection_pool();
+            }
+        }
+    }
+
+    /// Marks a server in the cluster as unknown due to the given `error`.
+    async fn mark_server_as_unknown(&self, error: Error, address: StreamAddress) {
+        let description = ServerDescription::new(address, Some(Err(error)));
+        self.update(description).await;
+    }
+
+    /// Updates the provided topology in a minimally contentious way by cloning first.
+    ///
+    /// Returns true if the topology changed as a result of the update and false otherwise.
+    pub(crate) async fn update(&self, server_description: ServerDescription) -> bool {
+        // Because we're calling clone on the lock guard, we're actually copying the TopologyState
+        // itself, not just making a new reference to it. The `servers` field will contain
+        // references to the same instances though, since each is wrapped in an `Arc`.
+        let mut state_clone = self.state.read().await.clone();
+
+        // TODO RUST-232: Theoretically, `TopologyDescription::update` can return an error. However,
+        // this can only happen if we try to access a field from the isMaster response when an error
+        // occurred during the check. In practice, this can't happen, because the SDAM algorithm
+        // doesn't check the fields of an Unknown server, and we only return Unknown server
+        // descriptions when errors occur. Once we implement SDAM monitoring, we can
+        // properly inform users of errors that occur here.
+        let _ = state_clone.update(server_description, self.clone());
+
+        // Now that we have the proper state in the copy, acquire a lock on the proper topology and
+        // move the info over.
+        let mut state_lock = self.state.write().await;
+        state_lock.description = state_clone.description;
+        state_lock.servers = state_clone.servers;
+
+        self.common.message_manager.notify_topology_changed();
```

Comment:
According to the SDAM spec, we're only supposed to publish topology description changed events if the topology description actually changed. I figure that also applies to server selection, since if the topology description didn't actually change, then selection will fail again. Given that, we should probably check for equality of the server/topology descriptions here before notifying that the topology changed. Also, there isn't really a point to returning a bool from this method if it will always be true.

Reply:
I agree that this should be updated, but when I realized that we hadn't done this already, I figured it would make sense to do in a follow-up commit rather than bundling the new equality implementation on `TopologyDescription` as part of this commit. Would you be okay with me adding a todo and filing a ticket for this to work on as soon as this is merged?

Reply:
Yep, sounds good to me.

Reply:
Filed RUST-301